src/hash/rpo/mod.rs

```diff
@@ -97,11 +97,7 @@ impl Hasher for Rpo256 {
         // compute the number of elements required to represent the string; we will be processing
         // the string in BINARY_CHUNK_SIZE-byte chunks, thus the number of elements will be equal
         // to the number of such chunks (including a potential partial chunk at the end).
-        let num_elements = if bytes.len() % BINARY_CHUNK_SIZE == 0 {
-            bytes.len() / BINARY_CHUNK_SIZE
-        } else {
-            bytes.len() / BINARY_CHUNK_SIZE + 1
-        };
+        let num_elements = bytes.len() / BINARY_CHUNK_SIZE;
```

Comment:
I'm confused by this and the below comment. I thought the first element of the capacity range is supposed to be either 1 or 0, not the number of elements to be hashed anymore. @Al-Kindi-0?

Reply:
We have indeed a couple of unanswered questions there. I was thinking of moving the PR to draft until we have these answered.

We indeed  set the first element as the count

https://github.com/0xPolygonMiden/crypto/blob/6de7730e30353c4c889591a52126415ac4819349/src/hash/rpo/mod.rs#L109-L110

But during the first iteration, we override it with the input, as `i` is set as `0`

https://github.com/0xPolygonMiden/crypto/blob/6de7730e30353c4c889591a52126415ac4819349/src/hash/rpo/mod.rs#L116-L136

There is a correct statement below that we don't need to add an extra element with `[1, 0, 0, 0, 0 ,0, 0]` because the number of elements in the initial inputs indeed will differentiate cases with or without remainder chunks.

https://github.com/0xPolygonMiden/crypto/blob/6de7730e30353c4c889591a52126415ac4819349/src/hash/rpo/mod.rs#L143-L149

But, if we change it to be either `0` or `1`, then we have to apply an extra padding element for such cases.

Reply:
Another question:

If we apply a montgomery reduction here (as opposed to `Felt::from_mont` that will just take the raw number)
https://github.com/0xPolygonMiden/crypto/blob/6de7730e30353c4c889591a52126415ac4819349/src/hash/rpo/mod.rs#L132-L135

Why do we limit the chunk size to `7` bytes, increasing the permutation count?

https://github.com/0xPolygonMiden/crypto/blob/6de7730e30353c4c889591a52126415ac4819349/src/hash/rpo/mod.rs#L40-L41

Reply:
@Al-Kindi-0 thanks for the clarification. As for the second item, I think its safer to go for 8 bytes with montgomery redcution as the reduction itself should be far cheaper than a permutation - and safer.

Regarding the padding, can you link the latest paper we should use as reference? I'll reproduce in the code the definition we have there.

Reply:
> @Al-Kindi-0 thanks for the clarification. As for the second item, I think its best to go for 8 bytes with montgomery redcution as the reduction itself should be far cheaper than a permutation - and safer. wdyt?
>
 
I am leaning also towards this option but maybe we should benchmark the two approaches to get a better idea.

> Regarding the padding, can you link the latest paper we should use as reference? I'll reproduce in the code the definition we have there.

https://eprint.iacr.org/2022/1577.pdf

Reply:
@Al-Kindi-0 it's clear what needs to be done. As pointed by @grjte , capacity will be either one or zero.

The last question I have is about one of the rules in the paper: `2.5 The zero-length input is not allowed`. Why is that? It's not immediately obvious because if we just permutate zeroes, we will also have a valid output (I think). And, considering sponge hashes are usually infallible, I guess we have to come to an output for this case.

Reply:
I think this is was mainly done in order to simplify (thinking about) the new padding rule. Hashing of only zeros is still allowed but the length then is not zero.